File: raspberry-pi/src/wordclock/word_clock.py
import random
from datetime import datetime
from gif import display_gif
from clock_display_hal import ClockDisplayHAL


class WordClock:
    COLORS = [
        (255, 0, 0),  # Red
        (0, 255, 0),  # Green
        (0, 0, 255),  # Blue
        (255, 255, 0),  # Yellow
        (255, 0, 255),  # Magenta
        (0, 255, 255),  # Cyan
        (255, 255, 255),  # White
        (165, 42, 42),  # Brown
    ]

    def __init__(self, clock_display_hal,gif_path):
        self.last_hour = -1
        self.all_last_highlighted_words = ""
        self.clock_display_hal = clock_display_hal
        self.gif_path=gif_path

    def highlight_word(self, word, color=(255, 255, 255)):
        if word in ClockDisplayHAL.WORDS_TO_LEDS:
            self.clock_display_hal.display_word(word, color)
        
    def get_random_color(self):
        return random.choice(WordClock.COLORS)

    def add_highlighted_word(self, word):
        self.highlight_word(word, self.get_random_color())
        self.all_last_highlighted_words += word

    def display_time(self):
        now = datetime.now()
        hour = now.hour
        minute = now.minute
        self.clock_display_hal.clear_pixels(show=False)
        if hour != self.last_hour and self.gif_path:
            if minute == 0:
                display_gif(self.gif_path,self.clock_display_hal)
                self.clock_display_hal.clear_pixels(show=False)
                self.last_hour = hour

        self.add_highlighted_word("IL")
        self.add_highlighted_word("EST")
        all_highlighted_words = "ILEST"

        # Handle hours
        if hour == 0 :
            self.add_highlighted_word("MINUIT")
            all_highlighted_words += "MINUIT"
        elif hour == 12:
            self.add_highlighted_word("MIDI")
            all_highlighted_words += "MIDI"
        else:
            hour = now.hour % 12
            self.add_highlighted_word(f"HEURE_{hour}")
            all_highlighted_words += f"HEURE_{hour}"

            h = "HEURES" if hour != 1 else "HEURE"
            self.add_highlighted_word(h)
            all_highlighted_words += h


        # Handle minutes
        if minute < 5:
            self.add_highlighted_word("PILE")
            all_highlighted_words += "PILE"
        elif minute < 10:
            self.add_highlighted_word("CINQ")
            all_highlighted_words += "CINQ"
        elif minute < 15:
            self.add_highlighted_word("DIX")
            all_highlighted_words += "DIX"
        elif minute < 20:
            self.add_highlighted_word("ET")
            self.add_highlighted_word("QUART")
            all_highlighted_words += "ETQUART"
        elif minute < 25:
            self.add_highlighted_word("VINGT")
            all_highlighted_words += "VINGT"
        elif minute < 30:
            self.add_highlighted_word("VINGTCINQ")
            all_highlighted_words += "VINGTCINQ"
        elif minute < 35:
            self.add_highlighted_word("ET")
            self.add_highlighted_word("DEMI")
            all_highlighted_words += "ETDEMI"
        elif minute < 40:
            self.add_highlighted_word("MOINS")
            self.add_highlighted_word("VINGT")
            self.add_highlighted_word("CINQ")
            all_highlighted_words += "MOINSVINGTCINQ"
        elif minute < 45:
            self.add_highlighted_word("MOINS")
            self.add_highlighted_word("VINGT")
            all_highlighted_words += "MOINSVINGT"
        elif minute < 50:
            self.add_highlighted_word("MOINS")
            self.add_highlighted_word("LE")
            self.add_highlighted_word("QUART")
            all_highlighted_words += "MOINSLEQUART"
        elif minute < 55:
            self.add_highlighted_word("MOINS")
            self.add_highlighted_word("DIX")
            all_highlighted_words += "MOINSDIX"
        else:
            self.add_highlighted_word("MOINS")
            self.add_highlighted_word("CINQ")
            all_highlighted_words += "MOINSCINQ"

        # Update display only if there are changes
        if self.all_last_highlighted_words != all_highlighted_words:
            self.clock_display_hal.show()
            self.all_last_highlighted_words = all_highlighted_words

File: esp/wordclock/src/WordClock.cpp
#include "WordClock.h"
#include "SerialHelper.h"

const uint32_t COLORS[] = {
    0xFF0000,
    0x00FF00,
    0x0000FF,
    0xFFFF00,
    0xFF00FF,
    0x00FFFF,
    0xFFFFFF,
    0xA52A2A};

WordClock::WordClock(ClockDisplayHAL *clockDisplayHAL, NetworkManager *networkManager, GifPlayer *gifPlayer)
    : clockDisplayHAL(clockDisplayHAL), networkManager(networkManager), gifPlayer(gifPlayer), lastHour(-1), allLastHighlightedWords(""), gifDownloaded(false) {}

void WordClock::setup()
{
    downloadGIF();
}

void WordClock::downloadGIF()
{
    if (!gifDownloaded)
    {
        const char *gifUrl = "https://raw.githubusercontent.com/johniak/word-clock/refs/heads/main/raspberry-pi/heart_art_small.gif";
        if (networkManager->downloadGIF(gifUrl))
        {
            uint8_t *gifBuffer = networkManager->getGifBuffer();
            size_t gifSize = networkManager->getGifBufferSize();
            if (gifSize > 0 && gifBuffer != nullptr)
            {
                if (gifPlayer->loadGIF(gifBuffer, gifSize))
                {
                    gifDownloaded = true;
                    SERIAL_PRINTLN("GIF downloaded and loaded successfully.");
                }
            }
        }
        else
        {
            SERIAL_PRINTLN("Failed to download GIF.");
        }
    }
}

void WordClock::highlightWord(const String &word, uint32_t color)
{
    clockDisplayHAL->displayWord(word, color);
}

uint32_t WordClock::getRandomColor()
{
    int index = random(0, sizeof(COLORS) / sizeof(COLORS[0]));
    return COLORS[index];
}



void WordClock::displayTime()
{
    struct tm currentTime = networkManager->getLocalTimeStruct();
    int hour = currentTime.tm_hour;
    int minute = currentTime.tm_min;

    showTime(hour, minute, true);
}

void WordClock::showTime(int hour, int minute, bool showRainbow)
{
    clockDisplayHAL->clearPixels(false);

    if (hour != lastHour && minute == 0 && showRainbow)
    {
        lastHour = hour;
        clockDisplayHAL->playRainbow(5000);
        if (gifDownloaded)
        {
            gifPlayer->playGIF(5000);
        }
        clockDisplayHAL->clearPixels(false);
    }

    highlightWord("IL", getRandomColor());
    highlightWord("EST", getRandomColor());
    String allHighlightedWords = "ILEST";

    // Handle hours
    if (minute >= 35)
    {
        hour = hour + 1;
    }
    if (hour==0 || hour==24) {
        highlightWord("MINUIT", getRandomColor());
        allHighlightedWords += "MINUIT";
    } else if (hour==12) {
        highlightWord("MIDI", getRandomColor());
        allHighlightedWords += "MIDI";
    } else {
        
        hour = hour % 12;

        String hourWord = "HEURE_" + String(hour);
        highlightWord(hourWord, getRandomColor());
        allHighlightedWords += hourWord;

        String h = (hour != 1) ? "HEURES" : "HEURE";
        highlightWord(h, getRandomColor());
        allHighlightedWords += h;
    }

    // Handle minutes
    if (minute < 5) {
        highlightWord("PILE", getRandomColor());
        allHighlightedWords += "PILE";
    } else if (minute < 10) {
        highlightWord("CINQ", getRandomColor());
        allHighlightedWords += "CINQ";
    } else if (minute < 15) {
        highlightWord("DIX", getRandomColor());
        allHighlightedWords += "DIX";
    } else if (minute < 20) {
        highlightWord("ET", getRandomColor());
        highlightWord("QUART", getRandomColor());
        allHighlightedWords += "ETQUART";
    } else if (minute < 25) {
        highlightWord("VINGT", getRandomColor());
        allHighlightedWords += "VINGT";
    } else if (minute < 30) {
        highlightWord("VINGT", getRandomColor());
        highlightWord("CINQ", getRandomColor());
        allHighlightedWords += "VINGTCINQ";
    }
    else if (minute < 35)
    {
        highlightWord("ET", getRandomColor());
        highlightWord("DEMI", getRandomColor());
        allHighlightedWords += "ETDEMI";
    }
    else if (minute < 40)
    {
        highlightWord("MOINS", getRandomColor());
        highlightWord("VINGT", getRandomColor());
        highlightWord("CINQ", getRandomColor());
        allHighlightedWords += "MOINSVINGTCINQ";
    }
    else if (minute < 45)
    {
        highlightWord("MOINS", getRandomColor());
        highlightWord("VINGT", getRandomColor());
        allHighlightedWords += "MOINSVINGT";
    }
    else if (minute < 50)
    {
        highlightWord("MOINS", getRandomColor());
        highlightWord("LE", getRandomColor());
        highlightWord("QUART", getRandomColor());
        allHighlightedWords += "MOINSLEQUART";
    }
    else if (minute < 55)
    {
        highlightWord("MOINS", getRandomColor());
        highlightWord("DIX", getRandomColor());
        allHighlightedWords += "MOINSDIX";
    }
    else
    {
        highlightWord("MOINS", getRandomColor());
        highlightWord("CINQ", getRandomColor());
        allHighlightedWords += "MOINSCINQ";
    }

    if (allLastHighlightedWords != allHighlightedWords)
    {
        clockDisplayHAL->show();
        allLastHighlightedWords = allHighlightedWords;
    }
}
